misk-core/src/main/kotlin/misk/logging/SampledLogger.kt

```diff
@@ -0,0 +1,245 @@
+package misk.logging
+
+import misk.sampling.Sampler
+import mu.KLogger
+import mu.Marker
+
+class SampledLogger constructor(
+  override val underlyingLogger: KLogger,
+  private val sampler: Sampler
+) : KLogger {
+
+  override fun <T : Throwable> catching(throwable: T) = if (sampler.sample()) underlyingLogger.catching(
+    throwable) else Unit
+
+  override fun debug(msg: () -> Any?) = if (sampler.sample()) underlyingLogger.debug(msg) else Unit
+
+  override fun debug(t: Throwable?, msg: () -> Any?)
+    = if (sampler.sample()) underlyingLogger.debug(t, msg) else Unit
```

Comment:
This is better, but I think we can make it even a bit cleaner with the help of a helper method! What do you think about something like this? We'll have to make `sampler` public which is fine.

```kotlin
  override fun debug(msg: () -> Any?) = sampled { underlyingLogger.debug(msg) }
  override fun debug(t: Throwable?, msg: () -> Any?) = sampled { underlyingLogger.debug(t, msg) }
  override fun debug(marker: Marker?, msg: () -> Any?) = sampled { underlyingLogger.debug(marker, msg) }

  inline fun sampled(f: () -> Unit) {
    if (sampler.sample()) f()
  }
```

Reply:
Whoa thanks, didn't know about `f()` in Kotlin. I think I can also make `sampled` a private function